src/zip/file.cr

```diff
@@ -0,0 +1,190 @@
+      i -= 1
+    end
+
+    i == -1 ? nil : size - buf_size + i
+  end
+
+  private def read_directory_end(directory_end_offset)
+    @io.pos = directory_end_offset
+
+    signature = read UInt32
+    if signature != Zip::END_OF_CENTRAL_DIRECTORY_HEADER_SIGNATURE
+      raise Error.new("expected end of central directory header signature, not 0x#{signature.to_s(16)}")
+    end
+
+    read Int16                     # number of this disk
+    read Int16                     # disk start
+    read Int16                     # number of entries in disk
+    entries_size = read Int16      # number of total entries
+    read UInt32                    # size of the central directory
+    directory_offset = read UInt32 # offset of central directory
+    comment_length = read UInt16   # comment length
+    if comment_length != 0
+      @comment = @io.read_string(comment_length)
+    end
+    {entries_size, directory_offset}
+  end
+
+  private def read_entries(directory_offset, entries_size)
+    @io.pos = directory_offset
+
+    entries_size.times do
+      signature = read UInt32
+      if signature != Zip::CENTRAL_DIRECTORY_HEADER_SIGNATURE
+        raise Error.new("expected directory header signature, not 0x#{signature.to_s(16)}")
+      end
+
+      entry = Entry.new(@io)
+      @entries << entry
+      @entries_by_filename[entry.filename] = entry
+    end
+  end
+
+  private def read(type)
+    @io.read_bytes(type, IO::ByteFormat::LittleEndian)
+  end
+
+  # An entry inside a `Zip::File`.
+  #
+  # Use the `open` method to read from it.
+  class Entry
+    include FileInfo
+
+    # :nodoc:
+    def initialize(@io : ::File | ::IO::Memory)
+      super(at_central_directory_header: io)
+    end
+
+    # Yields an `IO` to read this entry's contents.
+    # Multiple entries can be opened and read concurrently.
+    def open
+      @io.read_at(data_offset.to_i32, compressed_size.to_i32) do |io|
+        io = decompressor_for(io, is_sized: true)
+        checksum_reader = ChecksumReader.new(io, filename, verify: crc32)
+        yield checksum_reader
+      end
+    end
+
+    private getter(data_offset : UInt32) do
+      # Apparently a zip entry might have different extra bytes
+      # in the local file header and central directory header,
+      # so to know the data offset we must read them again.
+      @io.read_at(offset.to_i32, 30) do |io|
+        # at least check that the signature is OK
+        signature = read(io, UInt32)
+        if signature != FileInfo::SIGNATURE
+          raise Zip::Error.new("wrong local file header signature (expected 0x#{FileInfo::SIGNATURE.to_s(16)}, got 0x#{signature.to_s(16)})")
+        end
+
+        # Skip most of the headers except filename length and extra length
+        io.skip(22)
+        filename_length = read(io, UInt16)
+        extra_length = read(io, UInt16)
+        @offset + 30 + filename_length + extra_length
```

Comment:
It's basically skipping 22 bytes from the header instead of reading them one by one. 30 is the size of the header without counting the filename length and extra length.